pkg/api/api.go

```diff
@@ -67,6 +67,148 @@ import (
 // loggerName is the tree path name of the logger for this package.
 const loggerName = "api"
 
+type headerDescriptor struct {
+	// name of header parameter
+	Name string
+
+	// Validate will validate parameter value or
+	// return error if value is not in correct format.
+	Validate func(v interface{}, s *Service) error
+
+	// If this function is defined it will set value to request context
+	SetToContext func(ctx context.Context, parsedValue interface{}) context.Context
+}
+
+var (
+	swarmPinHeader = headerDescriptor{
+		Name: "Swarm-Pin",
+		Validate: func(v interface{}, _ *Service) error {
+			v = strings.ToLower(v.(string))
+			switch v {
+			case v.(*bool) == nil || !v.(bool) || v.(bool):
+				return nil
+			default:
+				return fmt.Errorf("%w: expected bool", errInvalidHeaderParams)
```

Comment:
What is the purpose of wrapping the `errInvalidHeaderParams`, where it is unwrapped? The same is below.

Reply:
sorry it's still a draft pr so i'm playing around with different things. Hopefully will have some proper implementation by tomorrow.